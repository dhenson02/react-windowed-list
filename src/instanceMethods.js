// external dependencies
import isFunction from 'lodash/isFunction';
import isUndefined from 'lodash/isUndefined';
import noop from 'lodash/noop';
import {
  findDOMNode
} from 'react-dom';

// constants
import {
  OFFSET_SIZE_KEYS,
  OVERFLOW_KEYS,
  SCROLL_START_KEYS,
  ADD_EVENT_LISTENER_OPTIONS,
  VALID_TYPES
} from './constants';

// utils
import {
  getCalculatedElementEnd,
  getCalculatedItemSizeAndItemsPerRow,
  getCalculatedSpaceBefore,
  getFromAndSize,
  getFromAndSizeFromListItemSize,
  getOffset,
  getScrollSize,
  getViewportSize,
  hasDeterminateSize,
  setCacheSizes,
  shouldSetState
} from './utils';

export const createGetDomNode = (instance) => {
  /**
   * @function getDomNode
   *
   * @description
   * get the DOM node of the property on the instance specified, or the instance itself
   *
   * @param {string} [property] the property on the instance requested
   * @returns {HTMLElement} the element requested
   */
  return (property) => {
    return findDOMNode(property ? instance[property] : instance);
  };
};

export const createGetItemSizeAndItemsPerRow = (instance) => {
  /**
   * @function getItemSizeAndItemsPerRow
   *
   * @description
   * get the itemSize and itemsPerRow values based on props
   *
   * @returns {{itemSize: number, itemsPerRow: number}} the itemSize and itemsPerRow
   */
  return () => {
    const {
      axis,
      useStaticSize
    } = instance.props;

    let {
      itemSize,
      itemsPerRow
    } = instance.state;

    if (useStaticSize && itemSize && itemsPerRow) {
      return {
        itemSize,
        itemsPerRow
      };
    }

    const itemElements = instance.getDomNode('items').children;

    return !itemElements.length ? {} : getCalculatedItemSizeAndItemsPerRow(itemElements, axis, itemSize);
  };
};

export const createGetScrollOffset = (instance) => {
  /**
   * @function getScrollOffset
   *
   * @description
   * get the scroll offset based on props
   *
   * @returns {number} the scrollOffset to apply
   */
  return () => {
    const {
      axis
    } = instance.props;

    const scrollKey = SCROLL_START_KEYS[axis];
    const actual = instance.scrollParent === window ?
      (document.body[scrollKey] || document.documentElement[scrollKey]) :
      instance.scrollParent[scrollKey];

    const max = getScrollSize(instance.scrollParent, axis) - getViewportSize(instance.scrollParent, axis);
    const scroll = Math.max(0, Math.min(actual, max));
    const element = instance.getDomNode();

    return getOffset(instance.scrollParent, axis) + scroll - getOffset(element, axis);
  };
};

export const createGetScrollParent = (instance) => {
  /**
   * @function getScrollParent
   *
   * @description
   * get the scroll parent element
   *
   * @returns {HTMLElement} the scroll parent
   */
  return () => {
    const {
      axis,
      scrollParentGetter
    } = instance.props;

    if (scrollParentGetter) {
      return scrollParentGetter();
    }

    const overflowKey = OVERFLOW_KEYS[axis];

    let element = findDOMNode(instance),
        overflowStyleKey;

    while (element = element.parentElement) {
      overflowStyleKey = window.getComputedStyle(element)[overflowKey];

      if (overflowStyleKey === 'auto' || overflowStyleKey === 'scroll' || overflowStyleKey === 'overlay') {
        return element;
      }
    }

    return window;
  };
};

export const createGetSizeOfListItem = (instance) => {
  /**
   * @function getSizeOfListItem
   *
   * @description
   * get the size of the list item requested
   *
   * @param {number} index the index of the list item
   * @returns {number} the size of the list item
   */
  return (index) => {
    const {
      axis,
      itemSizeGetter,
      itemSizeEstimator,
      type
    } = instance.props;
    const {
      from,
      itemSize,
      size
    } = instance.state;

    // Try the static itemSize.
    if (itemSize) {
      return itemSize;
    }

    // Try the itemSizeGetter.
    if (isFunction(itemSizeGetter)) {
      return itemSizeGetter(index);
    }

    // Try the cache.
    if (!isUndefined(instance.cache[index])) {
      return instance.cache[index];
    }

    // Try the DOM.
    if (type === VALID_TYPES.SIMPLE && index >= from && index < from + size && instance.items) {
      const itemEl = instance.getDomNode('items').children[index - from];

      if (itemEl) {
        return itemEl[OFFSET_SIZE_KEYS[axis]];
      }
    }

    // Try the itemSizeEstimator.
    if (itemSizeEstimator) {
      return itemSizeEstimator(index, instance.cache);
    }

    return 0;
  };
};

export const createGetSpaceBefore = (instance) => {
  /**
   * @function getSpaceBefore
   *
   * @description
   * get the space before the item requested
   *
   * @param {number} index the index of the item requested
   * @param {Object} [cache={}] the instance cache
   * @returns {number} the space before the item requested
   */
  return (index, cache = {}) => {
    if (!isUndefined(cache[index])) {
      return cache[index];
    }

    // Try the static itemSize.
    const {
      itemSize,
      itemsPerRow
    } = instance.state;

    cache[index] = itemSize ? Math.floor(index / itemsPerRow) * itemSize :
      getCalculatedSpaceBefore(cache, index, instance.getSizeOfListItem);

    return cache[index];
  };
};

export const createGetStartAndEnd = (instance) => {
  /**
   * @function getStartAndEnd
   *
   * @description
   * get the start and end values based on scroll position
   *
   * @param {number} [threshold=this.props.threshold] the pixel threshold to scroll above and below
   * @returns {{end: number, start: number}} the start and end of the window
   */
  return (threshold = instance.props.threshold) => {
    const {
      axis,
      itemSizeGetter,
      length,
      type
    } = instance.props;

    const scroll = instance.getScrollOffset();
    const start = Math.max(0, scroll - threshold);
    const calculatedEnd = scroll + getViewportSize(instance.scrollParent, axis) + threshold;
    const end = !hasDeterminateSize(type, itemSizeGetter) ? calculatedEnd :
      Math.min(calculatedEnd, instance.getSpaceBefore(length));

    return {
      end,
      start
    };
  };
};

export const createRenderItems = (instance) => {
  /**
   * @function renderItems
   *
   * @description
   * render the items that are currently visible
   *
   * @returns {ReactElement} the rendered container with the items
   */
  return () => {
    const {
      itemRenderer,
      itemsRenderer
    } = instance.props;
    const {
      from,
      size
    } = instance.state;

    const items = [];

    let index = -1;

    while (++index < size) {
      items[index] = itemRenderer(from + index, index);
    }

    return itemsRenderer(items, (containerRef) => {
      return instance.items = containerRef;
    });
  };
};

export const createScrollTo = (instance) => {
  /**
   * @function scrollTo
   *
   * @description
   * scroll the element to the requested initialIndex
   */
  return () => {
    const {
      initialIndex
    } = instance.props;

    if (!isUndefined(initialIndex)) {
      instance.setScroll(instance.getSpaceBefore(initialIndex));
    }
  };
};

export const createSetScroll = (instance) => {
  /**
   * @function setScroll
   *
   * @description
   * set the scroll based on the current offset
   *
   * @param {number} currentOffset the current offset
   * @returns {void}
   */
  return (currentOffset) => {
    if (!instance.scrollParent) {
      return;
    }

    const {
      axis
    } = instance.props;

    let offset = currentOffset + getOffset(instance.getDomNode(), axis);

    if (instance.scrollParent === window) {
      return window.scrollTo(0, offset);
    }

    offset -= getOffset(instance.scrollParent, axis);

    instance.scrollParent[SCROLL_START_KEYS[axis]] = offset;
  };
};

export const createSetStateIfAppropriate = (instance) => {
  /**
   * @function setStateIfAppropriate
   *
   * @description
   * set the state if shouldSetState returns true
   *
   * @param {Object} nextState the possible next state of the instance
   * @param {function} callback the callback to call once the state is set
   * @returns {void}
   */
  return (nextState, callback) => {
    if (shouldSetState(instance.state, nextState)) {
      return callback();
    }

    instance.setState(nextState, callback);
  };
};

export const createUpdateFrame = (instance) => {
  /**
   * @function updateFrame
   *
   * @description
   * update the frame based on the type in props
   *
   * @param {function} callback the function to call once the frame is updated
   * @returns {void}
   */
  return (callback) => {
    const {
      type
    } = instance.props;

    instance.updateScrollParent();

    if (!isFunction(callback)) {
      callback = noop;
    }

    if (type === VALID_TYPES.UNIFORM) {
      return instance.updateUniformFrame(callback);
    }

    if (type === VALID_TYPES.VARIABLE) {
      return instance.updateVariableFrame(callback);
    }

    return instance.updateSimpleFrame(callback);
  };
};

export const createUpdateScrollParent = (instance) => {
  /**
   * @function updateScrollParent
   *
   * @description
   * update the scroll parent with the listeners it needs
   *
   * @returns {void}
   */
  return () => {
    const newScrollParent = instance.getScrollParent();

    if (newScrollParent === instance.scrollParent) {
      return;
    }

    if (instance.scrollParent) {
      instance.scrollParent.removeEventListener('scroll', instance.updateFrame);
      instance.scrollParent.removeEventListener('mousewheel', noop);
    }

    instance.scrollParent = newScrollParent;

    instance.scrollParent.addEventListener('scroll', instance.updateFrame, ADD_EVENT_LISTENER_OPTIONS);
    instance.scrollParent.addEventListener('mousewheel', noop, ADD_EVENT_LISTENER_OPTIONS);
  };
};

export const createUpdateSimpleFrame = (instance) => {
  /**
   * @function updateSimpleFrame
   *
   * @description
   * update the frame when the type is 'simple'
   *
   * @param {function} callback the function to call once the frame is updated
   * @returns {void}
   */
  return (callback) => {
    const {
      end
    } = instance.getStartAndEnd();

    const itemElements = instance.getDomNode('items').children;
    const elementEnd = getCalculatedElementEnd(itemElements, instance.props);

    if (elementEnd > end) {
      return callback();
    }

    const {
      pageSize,
      length
    } = instance.props;

    const size = Math.min(instance.state.size + pageSize, length);

    instance.setStateIfAppropriate({
      size
    }, callback);
  };
};

export const createUpdateUniformFrame = (instance) => {
  /**
   * @function updateUniformFrame
   *
   * @description
   * update the frame when the type is 'uniform'
   *
   * @param {function} callback the function to call once the frame is updated
   * @returns {void}
   */
  return (callback) => {
    const itemSizeAndItemsPerRow = instance.getItemSizeAndItemsPerRow();
    const {
      itemSize,
      itemsPerRow
    } = itemSizeAndItemsPerRow;

    if (!itemSize || !itemsPerRow) {
      return callback();
    }

    const {
      start,
      end
    } = instance.getStartAndEnd();

    const calculatedFrom = Math.floor(start / itemSize) * itemsPerRow;
    const calulatedSize = (Math.ceil((end - start) / itemSize) + 1) * itemsPerRow;
    const fromAndSize = getFromAndSize(calculatedFrom, calulatedSize, itemsPerRow, instance.props);

    return instance.setStateIfAppropriate({
      ...fromAndSize,
      ...itemSizeAndItemsPerRow
    }, callback);
  };
};

export const createUpdateVariableFrame = (instance) => {
  /**
   * @function updateVariableFrame
   *
   * @description
   * update the frame when the type is 'variable'
   *
   * @param {function} callback the function to call once the frame is updated
   * @returns {void}
   */
  return (callback) => {
    const {
      axis,
      itemSizeGetter
    } = instance.props;
    const {
      from: currentFrom
    } = instance.state;

    if (!itemSizeGetter) {
      setCacheSizes(currentFrom, instance.getDomNode('items'), axis, instance.cache);
    }

    const startAndEnd = instance.getStartAndEnd();
    const fromAndSize = getFromAndSizeFromListItemSize(startAndEnd, instance.props, instance.getSizeOfListItem);

    instance.setStateIfAppropriate(fromAndSize, callback);
  };
};
