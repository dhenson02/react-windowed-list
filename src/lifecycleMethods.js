// external dependencies
import noop from 'lodash/noop';
import raf from 'raf';

// constants
import {
  MAX_SYNC_UPDATES,
  ADD_EVENT_LISTENER_OPTIONS,
  UNSTABLE_MESSAGE
} from './constants';

// utils
import {
  getFromAndSize
} from './utils';

export const createComponentDidMount = (instance) => {
  /**
   * @function componentDidMount
   *
   * @description
   * on mount, update the frame with the desired scroll position
   */
  return () => {
    raf(() => {
      instance.updateFrame(instance.scrollTo());
    });
  };
};

export const createComponentDidUpdate = (instance) => {
  /**
   * @function componentDidUpdate
   *
   * @description
   * update the frame position, cutting off after a certain point if unstable
   *
   * @returns {void}
   */
  return () => {
    if (instance.unstable) {
      return;
    }

    if (++instance.updateCounter > MAX_SYNC_UPDATES) {
      instance.unstable = true;

      /* eslint-disable no-console */
      return console.error(UNSTABLE_MESSAGE);
      /* eslint-enable */
    }

    if (!instance.updateCounterTimeoutId) {
      instance.updateCounterTimeoutId = raf(() => {
        instance.updateCounter = 0;

        delete instance.updateCounterTimeoutId;
      });
    }

    raf(() => {
      instance.updateFrame();
    });
  };
};

export const createComponentWillMount = (instance) => {
  /**
   * @function componentWillMount
   *
   * @description
   * prior to mount, set the initial state
   */
  return () => {
    const {
      initialIndex
    } = instance.props;
    const itemsPerRow = 1;
    const fromAndSize = getFromAndSize(initialIndex, 0, itemsPerRow, instance.props);

    instance.setState({
      ...fromAndSize,
      itemsPerRow
    });
  };
};

export const createComponentWillReceiveProps = (instance) => {
  /**
   * @function componentWillReceiveProps
   *
   * @description
   * when props are received, set the state if the next calculated state is different
   *
   * @param {Object} nextProps the incoming props
   */
  return (nextProps) => {
    const {
      from,
      itemsPerRow,
      size
    } = instance.state;

    instance.setStateIfAppropriate(getFromAndSize(from, size, itemsPerRow, nextProps), noop);
  };
};

export const createComponentWillUnmount = (instance) => {
  /**
   * @function componentWillUnmount
   *
   * @description
   * before unmount, remove any listeners applied to the scroll container
   */
  return () => {
    instance.scrollParent.removeEventListener('scroll', instance.updateFrame, ADD_EVENT_LISTENER_OPTIONS);
    instance.scrollParent.removeEventListener('mousewheel', noop, ADD_EVENT_LISTENER_OPTIONS);
  };
};
