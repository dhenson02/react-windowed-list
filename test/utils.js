// test
import test from 'ava';

// src
import * as utils from 'src/utils';

test.todo('defaultItemRenderer');
test.todo('defaultItemsRenderer');
test.todo('getOffset');
test.todo('getCalculatedElementEnd');
test.todo('getCalculatedSpaceBefore');
test.todo('getCalculatedItemSizeAndItemsPerRow');
test.todo('getContainerStyle');
test.todo('getFromAndSize');
test.todo('getFromAndSizeFromListItemSize');
test.todo('getListContainerStyle');
test.todo('getScrollSize');
test.todo('getViewportSize');
test.todo('hasDeterminateSize');
test.todo('renderItems');
test.todo('setCacheSizes');
test.todo('shouldSetState');
