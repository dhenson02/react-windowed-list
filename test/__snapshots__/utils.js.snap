// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`if defaultItemRenderer will render the item 1`] = `
<div>
  2
</div>
`;

exports[`if defaultContainerRenderer will render the item 1`] = `
<div>
  foo
  bar
  baz
</div>
`;
